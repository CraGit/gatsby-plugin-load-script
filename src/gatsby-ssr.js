import React from 'react';

export const onRenderBody = ({ setPostBodyComponents }, pluginOptions) => {
  const {
    plugins, // internal to gatsby
    disable, // should we disable the plugin ?
    src, // source of the script
    onerror,
    onError,
    onload,
    onLoad,
    async = true,
    charset = 'utf8',
    type = 'text/javascript',
    ...options
  } = pluginOptions;

  if (!src) {
    throw new Error(
      'gatsby-plugin-load-script needs a "src" option in order to work correctly',
    );
  }

  const finalOptions = {
    src,
    onerror: onError || onerror,
    onload: onLoad || onload,
    type,
    charset,
    async,
    ...options,
  };
  let optionsAsString = '';
  Object.entries(finalOptions).forEach(([property, value]) => {
    if (value !== undefined) {
      if (property === 'onerror' || property === 'onload') {
        optionsAsString += `script.${property}=${value};`;
      } else {
        optionsAsString += `script.${property}="${value}";`;
      }
    }
  });
  if (!disable)
    setPostBodyComponents([
      <script
        key={src}
        dangerouslySetInnerHTML={{
          __html: `
                        var head = document.head || document.getElementsByTagName('head')[0];
                        var script = document.createElement('script');
                        ${optionsAsString}
                        head.appendChild(script);
                    `,
        }}
      />,
    ]);
};
