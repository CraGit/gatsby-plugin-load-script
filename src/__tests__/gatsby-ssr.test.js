import {onRenderBody} from '../gatsby-ssr'

const setPostBodyComponents = jest.fn()

const simpleORB = (pluginOptions) =>
  onRenderBody({setPostBodyComponents}, {plugins: [], ...pluginOptions})

afterAll(() => {
  setPostBodyComponents.mockClear()
})

test('pluginsOptions need a "src" property', () => {
  expect(() => simpleORB({})).toThrowErrorMatchingSnapshot()
})

describe('onRenderBody should create post body components', () => {
  afterAll(() => {
    setPostBodyComponents.mockClear()
  })

  test('Disabled plugin should not add body components', () => {
    let SENTRY_DSN
    simpleORB({
      disable: !SENTRY_DSN,
      src: 'https://browser.sentry-cdn.com/5.4.0/bundle.min.js',
      onLoad: `() => Sentry.init({dsn:"${SENTRY_DSN}"})`,
    })
    expect(setPostBodyComponents.mock.calls.length).toBe(0)
  })

  test('Providing common Sentry options should add a postBodyComponent', () => {
    const SENTRY_DSN = 'this-is-a-testing-dsn'
    simpleORB({
      disable: !SENTRY_DSN,
      src: 'https://browser.sentry-cdn.com/5.4.0/bundle.min.js',
      onLoad: `() => Sentry.init({dsn:"${SENTRY_DSN}"})`,
    })
    expect(setPostBodyComponents.mock.calls.length).toBe(1)
    expect(setPostBodyComponents.mock.calls[0]).toMatchSnapshot()
  })
})
