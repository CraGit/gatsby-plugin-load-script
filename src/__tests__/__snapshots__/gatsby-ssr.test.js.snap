// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onRenderBody should create post body components Providing common Sentry options should add a postBodyComponent 1`] = `
Array [
  Array [
    <script
      dangerouslySetInnerHTML={
        Object {
          "__html": "var head = document.head || document.getElementsByTagName(\\"head\\")[0];
var script = document.createElement(\\"script\\");
script.src=\\"https://browser.sentry-cdn.com/5.4.0/bundle.min.js\\";
script.onload=() => Sentry.init({dsn:\\"this-is-a-testing-dsn\\"});
script.type=\\"text/javascript\\";
script.charset=\\"utf8\\";
script.async=true;
head.appendChild(script);",
        }
      }
    />,
  ],
]
`;

exports[`pluginsOptions need a "src" property 1`] = `"gatsby-plugin-load-script needs a \\"src\\" option in order to work correctly"`;
